feat: 일정 모달 Ui 구현

rkdcodus opened the pull request:
## 💡 작업 내용

- [x] 일정 추가 및 수정을 위한 모달 컴포넌트 UI 구현
- [x] 날짜 및 시간 선택 창 UI 구현
- [x] NormalTextarea 공통 컴포넌트 구현

## 💡 자세한 설명

### 1️⃣ `TaskModal` 모달 UI 구현 
Task 추가 및 수정을 위한 모달입니다. 
props 중 mode 를 통해 task 추가 모달, 수정 모달을 선택할 수 있습니다. 
수정일 경우에는 task prop으로 기존의 데이터를 전달합니다.

#### 사용 예시 
```tsx
<TaskModal
    mode={mode}
    isOpen={isOpen}
    setIsOpen={setIsOpen}
    task={null}
  />
```

### 2️⃣ `DayPickerModal` 모달 UI 구현
일정 모달에는 날짜와 시간을 입력하는 input이 있습니다. 
DayPickerModal은 input창을 클릭했을 때 날짜와 시간을 선택할 수 있는 또 다른 모달입니다.
현재는 연속일자는 선택 불가능합니다.

![image](https://github.com/user-attachments/assets/0c894a41-5ea0-4c7a-b32d-a451358d6f32)

기본적으로 현재 시각이 나오도록 구현했습니다. 
사용된 라이브러리는 `date-fns`와 `react-day-picker` 입니다. 선택은 npm trends를 기반으로 선택했습니다. 

DayPickerModal 컴포넌트에 작성된 아래 코드는 모달 외부를 클릭했을 때 DayPickerModal 모달이 닫히는 로직입니다. 

```ts
useEffect(() => {
    if (!isOpen) return;

    const outsideClick = (e: MouseEvent) => {
      if (pickerRef.current && !pickerRef.current.contains(e.target as Node)) {
        setIsOpen(false);
      }
    };

    document.addEventListener('mousedown', outsideClick);

    return () => document.removeEventListener('mousedown', outsideClick);
  }, [pickerRef, isOpen, setIsOpen]);
```

TaskModal은 해당 로직을 구현하지 않았습니다. 일정 작성 도중 닫히기 쉬운 구조는 사용자 입장에서 불편할 거라 생각하기 때문입니다.

### 3️⃣ NormalTextarea 공통 컴포넌트 
NormalInput과 마찬가지로 textarea 요소를 가공한 컴포넌트입니다.

아래 코드는 text를 길게 작성했을 때 textarea의 높이를 자동으로 조절하는 코드입니다!
```ts
  useEffect(() => {
    const ref = textareaRef.current;

    if (ref) {
      ref.style.height = 'auto';
      ref.style.height = ref.scrollHeight + 'px';
    }
  }, [text]);
```

### 4️⃣ 일정 모달의 추가 구현 사항
- 연속 일자 선택
- 위치 검색 

## 📗 참고 자료 (선택)

## 📢 리뷰 요구 사항 (선택)


## ✅ 셀프 체크리스트

- [x] 머지할 브랜치 확인했나요?
- [x] Reviewers, Labels, Projects를 등록했나요?
- [x] 기능이 잘 동작하나요?
- [x] 불필요한 코드는 제거했나요?

rkdcodus committed:
TTOL-14 chore: 달력 관련 라이브러리 설치

- date-fns
- react-day-picker
- react-big-calendar
- @types/react-big-calendar

rkdcodus committed:
TTOL-14 chore: png 파일 추가

- calendar-light
- close-white

rkdcodus committed:
TTOL-14 feat: NormalTextarea 컴포넌트 생성

rkdcodus committed:
TTOL-14 refactor: NormalInput read-only 스타일 추가

rkdcodus committed:
TTOL-14 feat: DayPickerModal 컴포넌트 생성

날짜 및 시간 선택 기능으로 `react-day-picker` 라이브러리 사용

rkdcodus committed:
TTOL-14 feat: TaskModal 컴포넌트 생성

일정 추가 및 수정을 위한 모달 컴포넌트

rkdcodus committed:
TTOL-14 style: textarea scroll 없애는 스타일변수 생성

rkdcodus committed:
TTOL-14 refactor: 날짜 및 시간 분리시킨 배열 변수명 수정

Jihye817 commented on src/app/_components/tasks/DayPickerModal.tsx:
```diff
@@ -0,0 +1,105 @@
+    onChange('start_time', undefined, formatISO(selected));
+  }, [selected, onChange]);
+
+  useEffect(() => {
```
DayPickerModal에만 외부 클릭 시 닫히는 로직 넣은 것 좋은 생각 같아요! 👍

Jihye817 commented on src/app/_components/tasks/DayPickerModal.tsx:
```diff
@@ -0,0 +1,105 @@
+  setIsOpen: Dispatch<SetStateAction<boolean>>;
+  onChange: (
+    field: keyof Task,
+    e?: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
```
모달 내에 TextArea가 없어서 Input Element 타입만 넣어도 될 것 같아요!

Jihye817 commented on src/app/_components/tasks/TaskModal.tsx:
```diff
@@ -0,0 +1,174 @@
+          <form className='mb-[20px] flex flex-col gap-[12px] rounded-[10px] bg-[#FAFAFA] p-2.5'>
+            <fieldset className='flex items-center gap-[20px] *:first:text-xl *:first:font-semibold *:last:flex-1'>
+              <label htmlFor='title'>제목</label>
+              <NormalInput
```
추후 일정 등록 시 사용자가 아무것도 입력하지 않아도 일정이 등록되는 경우를 방지하기 위해 input 태그 내부에 required 속성이 필요할 것 같습니다!

rkdcodus replied:
제목 required 속성 추가했습니다! 날짜는 기본값으로 현재 시각이 들어가고 메모와 위치는 선택 작성으로 남겨두었어요

Jihye817 commented on src/app/_components/common/NormalTextarea.tsx:
```diff
@@ -0,0 +1,57 @@
+    if (onChange) onChange(e);
+  };
+
+  useEffect(() => {
```
useEffect 대신에 함수로 정의 한 다음 onInput에서 실행되는 방식은 어떠신가요..? 프로젝트가 작아서 useEffect를 써도 상관없을 것 같긴 한데 바뀔 때 마다 렌더링이 일어나서 자원을 조금 더 소비한다는 말이 있더라구요!

```tsx
  const handleInput = () => {
    const ref = textareaRef.current;

    if (ref) {
      ref.style.height = 'auto';
      ref.style.height = ref.scrollHeight + 'px';
    }
  };

...

      <textarea
        {...rest}
        ref={textareaRef}
        onChange={handleChange}
        onInput={handleInput}
        className='scroll-none flex-1 resize-none outline-none'
      />
```

rkdcodus replied:
onInput을 많이 사용을 안해봐서 생각을 못했는데 의견 감사합니다!
찾아보니까 onInput을 사용하면 렌더링 될 필요 없이 즉각적으로 반응해서 높이 늘어나는게 조금이지만 더 빠르다고 하네요! 그렇지만 onInput을 사용하면 외부로부터 값이 변경될 때에는 onInput 트리거가 되지 않더라구요 둘 다 같이 사용하는 방법도 구현해봤는데 이렇게 하니까 높이 조절하는 코드가 한글자만 쳐도 여러번 호출되는 문제가 있었습니다 그래서 외부에서 value와 onChange를 관리하고 있기도 하고해서 useEffect 방식으로 가야할 것 같습니다..!
[669f6e9](https://github.com/kdt5-3rd/kdt5-3rd-front/pull/7/commits/669f6e99aeffbd079a588241566574ca7d79823f) 불필요한 코드는 삭제했습니다!

Jihye817 commented on src/app/_components/tasks/TaskModal.tsx:
```diff
@@ -0,0 +1,176 @@
+            <p className='text-3xl font-semibold'>{modalMode[mode].title}</p>
+            <button
+              onClick={handleCloseButton}
+              className='flex h-[30px] w-[30px] cursor-pointer items-center justify-center rounded-[10px] bg-[#7E95FF]'
```
피그마에 색상 토큰이 제대로 표시되지 않는 문제가 있나보네요🥲 #FFFFFF는 primary-0, #7E95FF는 primary-400 입니다! 이 부분은 제가 피그마 색상 토큰 설정을 다시 해보겠습니다..!

rkdcodus replied:
[6a84761](https://github.com/kdt5-3rd/kdt5-3rd-front/pull/7/commits/6a84761c496502b8a3f7263a4b37d786c9a8c724) 아직 적용 못했던 부분도 색상 토큰 적용했습니다!

Jihye817 reviewed (approved):
밤늦게까지 너무 수고하셨습니다!! ☺️
질문과 의견 달아두었어요!

rkdcodus committed:
TTOL-14 refactor: DayPickerModal onChange prop 타입 수정

rkdcodus committed:
TTOL-14 refactor: TaskModal title required 속성 추가

rkdcodus committed:
TTOL-14 refactor: NormalTextarea onInput 사용하여 높이 조절하도록 수정

하지만 이 경우 외부에서 값이 할당되었을 경우엔 조절되지 않기때문에 추가로 외부로부터 value 값이 있을 때 useEffect를 통해 높이 조절되도록 함.

rkdcodus committed:
TTOL-14 style: 색상 토큰 적용

rkdcodus committed:
TTOL-14 refactor: NormalTextarea 높이 자동 조절 useEffect 방식으로 다시 수정

Jihye817 reviewed (approved):
수고하셨습니다 🥰👍

rkdcodus committed:
Merge branch 'dev' of https://github.com/JungDoo/kdt5-3rd-front into TTOL-14-feature-modal-ui

rkdcodus committed:
TTOL-14 refactor: TaskCalendar 타입으로 수정
